File: leetcode/medium/MinimumPathSum.py
from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        """
        Finds the minimum path sum from the top-left to the bottom-right corner of a grid.
        You can only move either down or right at any point in time.

        This solution uses a bottom-up dynamic programming approach, modifying the grid
        in-place to store the minimum path sum to reach each cell. This results in
        O(1) extra space complexity.

        Algorithm:
        1. The value `grid[i][j]` will be updated to represent the minimum path sum
           to reach the cell at `(i, j)`.
        2. Initialize the first row: The path to any cell in the first row can only
           come from the left. So, `grid[0][j] += grid[0][j-1]`.
        3. Initialize the first column: The path to any cell in the first column can
           only come from above. So, `grid[i][0] += grid[i-1][0]`.
        4. For the rest of the cells `(i, j)`, the minimum path must come from either
           the cell above (`grid[i-1][j]`) or the cell to the left (`grid[i][j-1]`).
           We update `grid[i][j]` with its own value plus the minimum of these two.
        5. The final answer is the value in the bottom-right corner, `grid[m-1][n-1]`.

        Time Complexity: O(m * n), where m and n are the dimensions of the grid.
        Space Complexity: O(1), as we modify the grid in-place.
        """
        if not grid or not grid[0]:
            return 0

        m, n = len(grid), len(grid[0])

        # Initialize the first row (paths can only come from the left).
        for j in range(1, n):
            grid[0][j] += grid[0][j - 1]

        # Initialize the first column (paths can only come from above).
        for i in range(1, m):
            grid[i][0] += grid[i - 1][0]

        # Fill in the rest of the grid using the DP relation.
        for i in range(1, m):
            for j in range(1, n):
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1])

        # The bottom-right corner now holds the minimum path sum.
        return grid[-1][-1]

File: leetcode/medium/MinimumPathSum.java
/**
 * Solution for the Minimum Path Sum problem.
 * This class uses Dynamic Programming to find the path with the minimum sum
 * from the top-left corner to the bottom-right corner of a grid,
 * where movement is restricted to only down or right.
 * The solution modifies the input grid in-place to store the minimum path sums,
 * achieving O(1) extra space complexity.
 */
class Solution {
    /**
     * Finds the minimum path sum from the top-left (0, 0) to the bottom-right (m-1, n-1)
     * of the grid.
     *
     * @param grid The m x n grid containing non-negative integers. It is modified in-place.
     * @return The minimum path sum. Returns 0 if the grid is null or empty.
     */
    public int minPathSum(int[][] grid) {
        // 1. Handle edge case: null or empty grid
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }

        int m = grid.length;
        int n = grid[0].length;

        // The core idea is Dynamic Programming:
        // grid[i][j] will store the minimum path sum from (i, j) to the bottom-right.
        // DP relation: grid[i][j] = grid[i][j] + min(grid[i+1][j], grid[i][j+1])
        // We compute this from bottom-right towards top-left.

        // 2. Initialize the bottom-right cell (Base Case) - No change needed here, it's the start/end point.

        // 3. Initialize the last column (m-2 to 0)
        // grid[i][n-1] can only come from grid[i+1][n-1]
        for (int i = m - 2; i >= 0; i--) {
            grid[i][n - 1] += grid[i + 1][n - 1];
        }

        // 4. Initialize the last row (n-2 to 0)
        // grid[m-1][j] can only come from grid[m-1][j+1]
        for (int j = n - 2; j >= 0; j--) {
            grid[m - 1][j] += grid[m - 1][j + 1];
        }

        // 5. Fill the rest of the grid (m-2, n-2 to 0, 0)
        // Apply the general DP relation: grid[i][j] = current_value + min(down_path, right_path)
        for (int i = m - 2; i >= 0; i--) {
            for (int j = n - 2; j >= 0; j--) {
                grid[i][j] += Math.min(grid[i + 1][j], grid[i][j + 1]);
            }
        }

        // 6. The minimum path sum from (0, 0) to (m-1, n-1) is now stored in grid[0][0]
        return grid[0][0];
    }
}